onnxruntime/contrib_ops/cuda/bert/mask_index_impl.cu

```diff
@@ -0,0 +1,118 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "mask_index_impl.h"
+#include "core/providers/cuda/shared_inc/cuda_call.h"
+#include <cub/cub.cuh>
+
+using namespace cub;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+template <typename T, unsigned TPB>
+__global__ void MaskIndexKernelSmall(int sequence_length, const T* mask, int* mask_index) {
+  using BlockReduce = cub::BlockReduce<int, TPB>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  // blockIdx.x is b
+  const int offset = blockIdx.x * sequence_length;  // batch strides of sequence_length
+
+  cub::Min min;
+  int thread_data(sequence_length);
+
+  const int idx = offset + threadIdx.x;
+  if (threadIdx.x < sequence_length) {
+    const T val = mask[idx];
+    if (val == 0)  // masked position: report thread idx
+    {
+      thread_data = threadIdx.x;
+    }
+  }
+
+  const auto min_index = BlockReduce(temp_storage).Reduce(thread_data, min);
+
+  if (threadIdx.x == 0) {
+    mask_index[blockIdx.x] = min_index;
+  }
+}
+
+template <typename T, unsigned TPB>
+__global__ void MaskIndexKernel(int sequence_length, const T* mask, int* mask_index) {
+  using BlockReduce = cub::BlockReduce<int, TPB>;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  // blockIdx.x is b
+  const int offset = blockIdx.x * sequence_length;  // batch strides of sequence_length
+
+  cub::Min min;
+  int thread_data(sequence_length);
+
+  for (int i = threadIdx.x; i < sequence_length; i += TPB) {
+    const int idx = offset + i;
+    const T val = mask[idx];
+    if (val == 0)
+    {
+      thread_data = min(thread_data, i);
+    }
+  }
+
+  const auto min_index = BlockReduce(temp_storage).Reduce(thread_data, min);
+
+  if (threadIdx.x == 0) {
+    mask_index[blockIdx.x] = min_index;
+  }
+}
+
+template <typename T>
+bool LaunchMaskIndexKernel(
+    cudaStream_t stream,
+    const T* mask,
+    int* mask_index,
+    int batch_size,
+    int sequence_length)
+{
+  // Assume input mask total elements n = batch_size x sequence_length.
+  // mask_index is of length batch_size, and the value is the index of first mask (0) within each batch.
+  if (sequence_length <= 32) {
+    MaskIndexKernelSmall<T, 32><<<batch_size, 32, 0, stream>>>(sequence_length, mask, mask_index);
+  } else if (sequence_length <= 128) {
+    MaskIndexKernelSmall<T, 128><<<batch_size, 128, 0, stream>>>(sequence_length, mask, mask_index);
+  } else if (sequence_length == 384) {
+    MaskIndexKernelSmall<T, 384><<<batch_size, 384, 0, stream>>>(sequence_length, mask, mask_index);
+  } else {
+    MaskIndexKernel<T, 256><<<batch_size, 256, 0, stream>>>(sequence_length, mask, mask_index);
+  }
+
+  return CUDA_CALL(cudaPeekAtLastError());
```

Comment:
>cudaPeekAtLastError [](start = 19, length = 19)

Note that because cuda is async, this function may also return error codes from previous, asynchronous launches. It's meaningless to have this kind of code here.